![Ironhack logo](https://i.imgur.com/1QgrNNw.png)

# Lab | Data Cleaning

## Introduction

We keep seeing a common phrase that 80% of the work of a data scientist is data cleaning. We have no idea whether this number is accurate but a data scientist indeed spends lots of time and effort in collecting, cleaning and preparing the data for analysis. This is because datasets are usually messy and complex in nature. It is a very important ability for a data scientist to refine and restructure datasets into a usable state in order to proceed to the data analysis stage.

In this exercise, you will both practice the data cleaning techniques we discussed in the lesson and learn new techniques by looking up documentations and references. You will work on your own but remember the teaching staff is at your service whenever you encounter problems.

## Getting Started

Read the instructions for each cell and provide your answers. Make sure to test your answers in each cell and save. Jupyter Notebook should automatically save your work progress. But it's a good idea to periodically save your work manually just in case.


## Resources

[Data Cleaning Tutorial](https://www.tutorialspoint.com/python/python_data_cleansing.html)

[Data Cleaning with Numpy and Pandas](https://realpython.com/python-data-cleaning-numpy-pandas/#python-data-cleaning-recap-and-resources)

[Data Cleaning Video](https://www.youtube.com/watch?v=ZOX18HfLHGQ)

[Data Preparation](https://www.kdnuggets.com/2017/06/7-steps-mastering-data-preparation-python.html)

[Google Search](https://www.google.es/search?q=how+to+clean+data+with+python)

# Data Cleaning 

In [1]:
import pandas as pd

# Read the users dataset.

Take a look at what is the `users.csv` separator.

In [2]:
users = pd.read_csv('../data/users.csv', sep = '#')
users.shape

(40503, 14)

## Check its shape

See the number of rows and columns you're dealing.

## Use the .head() to see some rows of your dataframe.

In [3]:
users.head()

Id  Reputation         CreationDate   DisplayName       LastAccessDate  \
0  -1           1  2010-07-19 06:55:26     Community  2010-07-19 06:55:26   
1   2         101  2010-07-19 14:01:36  Geoff Dalgas  2013-11-12 22:07:23   
2   3         101  2010-07-19 15:34:50  Jarrod Dixon  2014-08-08 06:42:58   
3   4         101  2010-07-19 19:03:27        Emmett  2014-01-02 09:31:02   
4   5        6792  2010-07-19 19:03:57         Shane  2014-08-13 00:23:47   

                       WebsiteUrl            Location  \
0  http://meta.stackexchange.com/  on the server farm   
1        http://stackoverflow.com       Corvallis, OR   
2        http://stackoverflow.com        New York, NY   
3    http://minesweeperonline.com   San Francisco, CA   
4         http://www.statalgo.com        New York, NY   

                                             AboutMe  Views  UpVotes  \
0  <p>Hi, I'm not really a person.</p>\n\n<p>I'm ...      0     5007   
1  <p>Developer on the StackOverflow team.  Find ...     25        3   
2  <p><a href="http://blog.stackoverflow.com/2009...     22       19   
3  <p>currently at a startup in SF</p>\n\n<p>form...     11        0   
4  <p>Quantitative researcher focusing on statist...   1145      662   

   DownVotes  AccountId   Age                     ProfileImageUrl  
0       1920         -1   NaN                                 NaN  
1          0          2  37.0                                 NaN  
2          0          3  35.0                                 NaN  
3          0       1998  28.0  http://i.stack.imgur.com/d1oHX.jpg  
4          5      54503  35.0                                 NaN

## Get the data info. 

Which columns have a great number of missing values? How many space does this dataframe is occupying in your memory?

Expected output:
````
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40503 entries, 0 to 40502
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               40503 non-null  int64  
 1   Reputation       40503 non-null  int64  
 2   CreationDate     40503 non-null  object 
 3   DisplayName      40497 non-null  object 
 4   LastAccessDate   40503 non-null  object 
 5   WebsiteUrl       8158 non-null   object 
 6   Location         11731 non-null  object 
 7   AboutMe          9424 non-null   object 
 8   Views            40503 non-null  int64  
 9   UpVotes          40503 non-null  int64  
 10  DownVotes        40503 non-null  int64  
 11  AccountId        40503 non-null  int64  
 12  Age              8352 non-null   float64
 13  ProfileImageUrl  16540 non-null  object 
dtypes: float64(1), int64(6), object(7)
memory usage: 4.3+ MB
````

In [4]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40503 entries, 0 to 40502
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Id               40503 non-null  int64  
 1   Reputation       40503 non-null  int64  
 2   CreationDate     40503 non-null  object 
 3   DisplayName      40497 non-null  object 
 4   LastAccessDate   40503 non-null  object 
 5   WebsiteUrl       8158 non-null   object 
 6   Location         11731 non-null  object 
 7   AboutMe          9424 non-null   object 
 8   Views            40503 non-null  int64  
 9   UpVotes          40503 non-null  int64  
 10  DownVotes        40503 non-null  int64  
 11  AccountId        40503 non-null  int64  
 12  Age              8352 non-null   float64
 13  ProfileImageUrl  16540 non-null  object 
dtypes: float64(1), int64(6), object(7)
memory usage: 4.3+ MB


## Rename Id column to user_id.

Remember to store you results back at the dataframe.

In [5]:
users = users.rename(columns = {'Id' : 'user_id'})

# Import the `posts.csv` dataset.

Note that this is a `gzip compressed csv`. In order to read this file correctly, you'll have to read the documentation (or help) of your `pd.read_csv()` function and check the `compression` argument. Try to understand which value of `compression=...` you should put in order to read your dataframe. 

In [6]:
posts = pd.read_csv('../data/posts.csv.gzip', compression = 'gzip')

## Perform the same as above to understand a bit of your data (head, info, shape)

In [7]:
posts.head()

Id  PostTypeId  AcceptedAnswerId          CreaionDate  Score  ViewCount  \
0   1           1              15.0  2010-07-19 19:12:12     23     1278.0   
1   2           1              59.0  2010-07-19 19:12:57     22     8198.0   
2   3           1               5.0  2010-07-19 19:13:28     54     3613.0   
3   4           1             135.0  2010-07-19 19:13:31     13     5224.0   
4   5           2               NaN  2010-07-19 19:14:43     81        NaN   

                                                Body  OwnerUserId  \
0  <p>How should I elicit prior distributions fro...          8.0   
1  <p>In many different statistical methods there...         24.0   
2  <p>What are some valuable Statistical Analysis...         18.0   
3  <p>I have two groups of data.  Each with a dif...         23.0   
4  <p>The R-project</p>\n\n<p><a href="http://www...         23.0   

       LasActivityDate                                              Title  \
0  2010-09-15 21:08:26                      Eliciting priors from experts   
1  2012-11-12 09:21:54                                 What is normality?   
2  2013-05-27 14:48:36  What are some valuable Statistical Analysis op...   
3  2010-09-08 03:00:19  Assessing the significance of differences in d...   
4  2010-07-19 19:21:15                                                NaN   

   ... AnswerCount  CommentCount  FavoriteCount  LastEditorUserId  \
0  ...         5.0             1           14.0               NaN   
1  ...         7.0             1            8.0              88.0   
2  ...        19.0             4           36.0             183.0   
3  ...         5.0             2            2.0               NaN   
4  ...         NaN             3            NaN              23.0   

          LastEditDate   CommunityOwnedDate ParentId  ClosedDate  \
0                  NaN                  NaN      NaN         NaN   
1  2010-08-07 17:56:44                  NaN      NaN         NaN   
2  2011-02-12 05:50:03  2010-07-19 19:13:28      NaN         NaN   
3                  NaN                  NaN      NaN         NaN   
4  2010-07-19 19:21:15  2010-07-19 19:14:43      3.0         NaN   

  OwnerDisplayName LastEditorDisplayName  
0              NaN                   NaN  
1              NaN                   NaN  
2              NaN                   NaN  
3              NaN                   NaN  
4              NaN                   NaN  

[5 rows x 21 columns]

In [8]:
posts.shape

(91976, 21)

In [9]:
posts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91976 entries, 0 to 91975
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Id                     91976 non-null  int64  
 1   PostTypeId             91976 non-null  int64  
 2   AcceptedAnswerId       14700 non-null  float64
 3   CreaionDate            91976 non-null  object 
 4   Score                  91976 non-null  int64  
 5   ViewCount              42921 non-null  float64
 6   Body                   91756 non-null  object 
 7   OwnerUserId            90584 non-null  float64
 8   LasActivityDate        91976 non-null  object 
 9   Title                  42921 non-null  object 
 10  Tags                   42921 non-null  object 
 11  AnswerCount            42921 non-null  float64
 12  CommentCount           91976 non-null  int64  
 13  FavoriteCount          13246 non-null  float64
 14  LastEditorUserId       44611 non-null  float64
 15  La

## Rename Id column to post_id and OwnerUserId to user_id.

Again, remember to check that your results are correctly stored inside the dataframe.

In [10]:
posts = posts.rename(columns = {'Id' : 'post_id', 'OwnerUserId' : 'user_id'})

## Define new dataframes for users and posts with the following selected columns:

**users columns**: user_id, Reputation, Views, UpVotes, DownVotes  
**posts columns**: post_id, Score, user_id, ViewCount, CommentCount, Body

In [11]:
users = users.loc[:, ['user_id', 'Reputation', 'Views', 'UpVotes', 'DownVotes']]
posts = posts.loc[:, ['post_id', 'Score', 'user_id', 'ViewCount', 'CommentCount', 'Body']]

**Note:** Check the new posts dataframe's info. What is the most noticeable change? 

Explain why we have chosen only some columns of it in terms of efficiency.

In [12]:
posts.info()
# The new dataframe is less than a third of the original.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91976 entries, 0 to 91975
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   post_id       91976 non-null  int64  
 1   Score         91976 non-null  int64  
 2   user_id       90584 non-null  float64
 3   ViewCount     42921 non-null  float64
 4   CommentCount  91976 non-null  int64  
 5   Body          91756 non-null  object 
dtypes: float64(2), int64(3), object(1)
memory usage: 4.2+ MB


# Merge the new dataframes you have created, of users and posts. Create a dataframe called `posts_from_users`

You will need to make an inner [merge](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) of posts and users dataframes. 

Think carefully which should be the key(s) for your merging.

In [13]:
posts_from_users = pd.merge(left = posts, right = users, how = 'inner', on = 'user_id' )

## Check the number of duplicated rows.

Remember you can sum the results of a mask to get how many numbers the True value appeared in the results. This occurs because `True` is interpreted as `1` in Python whereas `False` is interpreted as `0`.

In [14]:
posts_from_users.duplicated().sum()

299

## Find those duplicate values and try to understand what happened.

*Hint:* You can use the argument `keep=False` from the `.duplicated()` method to bring the duplication.

*Hint 2:* You can sort the values `by=['user_id', 'post_id']` to see them in order.


In [15]:
posts.duplicated().sum()

0

In [16]:
users.duplicated().sum()
#There are duplicated rows in the users dataframe.

178

In [17]:
mask = posts_from_users.duplicated(keep = False)
posts_from_users.loc[mask,:].sort_values(by = ['user_id', 'post_id'])
#There were duplicated rows in users df, when merging with posts(inner) on 'user_id', generated duplicates row for
#every post made by those users.

post_id  Score  user_id  ViewCount  CommentCount  \
7481      1289      7    760.0     1139.0             8   
7482      1289      7    760.0     1139.0             8   
7483      8625      6    760.0     1799.0             3   
7484      8625      6    760.0     1799.0             3   
7485     23987      0    760.0       62.0             3   
...        ...    ...      ...        ...           ...   
90221   114527      0  54711.0       45.0             5   
90245   113334      3  54741.0      122.0             9   
90246   113334      3  54741.0      122.0             9   
90364   113691      0  54911.0       36.0            11   
90365   113691      0  54911.0       36.0            11   

                                                    Body  Reputation  Views  \
7481   <p>I am having difficulties to select the righ...         168     13   
7482   <p>I am having difficulties to select the righ...         168     13   
7483   <p>I was fiddling with PCA and LDA methods and...         168     13   
7484   <p>I was fiddling with PCA and LDA methods and...         168     13   
7485   <p>I was studying on a PAMI article and I have...         168     13   
...                                                  ...         ...    ...   
90221  <p>From Shapiro-Wilk's test I see that the res...           4     18   
90245  <p>I am confused on what I have read about the...          16      1   
90246  <p>I am confused on what I have read about the...          16      1   
90364  <p>I extract data related to a movie by sentim...           1      1   
90365  <p>I extract data related to a movie by sentim...           1      1   

       UpVotes  DownVotes  
7481        13          0  
7482        13          0  
7483        13          0  
7484        13          0  
7485        13          0  
...        ...        ...  
90221        0          0  
90245        0          0  
90246        0          0  
90364        0          0  
90365        0          0  

[598 rows x 10 columns]

## Should you drop it? If you think it is reasonable to drop it, then drop it.

Think: How would you correct it in the first place? That is, what was wrong in the first place?

*Hint:* There's a pandas method to drop duplicates. If you wanted to do it by hand, you could select the indexes of the duplicated values and `.drop()` it. 

In [18]:
#Drop the duplicated rows in dataframe users before merging with posts dataframe.
users1 = users.drop_duplicates()
#checking if it worked
users1.duplicated().sum()

0

In [19]:
#Try to merge the new users1 data frame with posts
posts_from_users1 = pd.merge(left = posts, right = users1, on = 'user_id')
#Check if there are duplicates
posts_from_users1.duplicated().sum()

0

In [20]:
#dropping duplicates manually
msk = posts_from_users.duplicated()
duplicated_index = posts_from_users.loc[msk,:].index
posts_from_users = posts_from_users.drop(index = duplicated_index)

In [21]:
#Checking if it worked
posts_from_users.duplicated().sum()

0

In [22]:
#dropping duplicates with pandas method
posts_from_users = posts_from_users.drop_duplicates()

In [23]:
#Checking if it worked
posts_from_users.duplicated().sum()

0

In [24]:
#Check if dataframes generated without the duplicates by the two methods are equal
posts_from_users.shape, posts_from_users1.shape

((90584, 10), (90584, 10))

## 10. How many missing values do you have in your merged dataframe? On which columns?

In [25]:
posts_from_users.isna().sum()

post_id             0
Score               0
user_id             0
ViewCount       48396
CommentCount        0
Body              220
Reputation          0
Views               0
UpVotes             0
DownVotes           0
dtype: int64

## Select only the rows in which there at least some missing values.

In [26]:
missing = posts_from_users.isna().any(axis = 1)
posts_from_users.loc[missing, :]

post_id  Score  user_id  ViewCount  CommentCount  \
1           16     16      8.0        NaN             3   
3           65     14      8.0        NaN             3   
4           78     33      8.0        NaN             4   
5          111      8      8.0        NaN             3   
7          129      4      8.0        NaN             0   
...        ...    ...      ...        ...           ...   
90843   115233      0  55599.0        NaN             0   
90844   115238      2  55599.0        NaN             0   
90856   115292      2  48622.0        NaN             0   
90857   115295      0  53248.0        NaN             0   
90868   115338      0  55729.0        NaN             0   

                                                    Body  Reputation  Views  \
1      <p>Two projects spring to mind:</p>\n\n<ol>\n<...        6764   1089   
3      <p>The first formula is the <em>population</em...        6764   1089   
4      <p>You tend to use the covariance matrix when ...        6764   1089   
5      <p>In R, the default setting for random number...        6764   1089   
7      <p>I quite like <a href="http://rads.stackover...        6764   1089   
...                                                  ...         ...    ...   
90843  <p>Before computing the variance-covariance ma...          31      2   
90844  <p>I agree with Glen_b. In regression problems...          31      2   
90856  <p>I don't think that a list with pros and con...          21      0   
90857  <p>You want to know what the probability is th...         101      0   
90868  <p>All items are not solved fully correctly in...          16      1   

       UpVotes  DownVotes  
1          604         25  
3          604         25  
4          604         25  
5          604         25  
7          604         25  
...        ...        ...  
90843        0          0  
90844        0          0  
90856        1          0  
90857        0          0  
90868        0          0  

[48396 rows x 10 columns]

## You will need to make something with missing values.  Will you clean or filling them? 

Pay attention. There can be different reasons for the missings numbers. Look at the `user_id` of some of them, look at the body of the message. Which ones you're sure of what should be and which one can you infer? Don't hurry up, take a look at your data.

In [23]:
missing_body = posts_from_users['Body'].isna()
posts_from_users.loc[missing_body, :]

post_id  Score  user_id  ViewCount  CommentCount Body  Reputation  \
689       3012      0     88.0        NaN             0  NaN       14082   
695       3018      0     88.0        NaN             0  NaN       14082   
817      13641      0     88.0        NaN             0  NaN       14082   
6589     56646      0    686.0        NaN             0  NaN       44152   
6604     57490      0    686.0        NaN             0  NaN       44152   
...        ...    ...      ...        ...           ...  ...         ...   
29793    72983      0   7290.0        NaN             0  NaN       37083   
29808    76603      0   7290.0        NaN             0  NaN       37083   
29810    76631      0   7290.0        NaN             0  NaN       37083   
29890    90803      0   7290.0        NaN             0  NaN       37083   
29919    93185      0   7290.0        NaN             0  NaN       37083   

       Views  UpVotes  DownVotes  
689     3320     4235        126  
695     3320     4235        126  
817     3320     4235        126  
6589    7357     2156         82  
6604    7357     2156         82  
...      ...      ...        ...  
29793   5554     8641        125  
29808   5554     8641        125  
29810   5554     8641        125  
29890   5554     8641        125  
29919   5554     8641        125  

[220 rows x 10 columns]

In [28]:
posts_from_users.loc[missing_body, :].Score.sum()

0

posts_from_users.loc[missing_body, :].CommentCount.sum()

In [29]:
posts_from_users.loc[missing_body, :].ViewCount.sum()

0.0

In [30]:
posts_from_users.loc[missing_body, :].CommentCount.sum()

0

Analizing the posts dataframe the features post_id and user_id as the name says its related to the identitity of the user and post, the other features(ViewCount, CommentCount, Score and Body) are directly related to the post made. So we can infer that when those four features are are 0 or NaN those row do not bring any useful data for analisys therefore we drop them.

In [24]:
posts_from_users = posts_from_users.drop(index = posts_from_users.loc[missing_body, :].index)

In [27]:
missing_viewcount = posts_from_users['ViewCount'].isna()
posts_from_users.loc[missing_viewcount, :].sort_values(by = 'CommentCount').tail(10)

post_id  Score  user_id  ViewCount  CommentCount  \
9986     26457     44    887.0        NaN            30   
29484    33781     69   7290.0        NaN            30   
68668    69462      0  28746.0        NaN            30   
17994     9636      3   2392.0        NaN            31   
64739    88453      5  25936.0        NaN            34   
23134    30160     12   4856.0        NaN            35   
7688      6605     30    795.0        NaN            37   
12835     2365      5   1124.0        NaN            41   
78546    92246     -5  38102.0        NaN            41   
43711    31038      6  11032.0        NaN            45   

                                                    Body  Reputation  Views  \
9986   <p>Part of the issue is that the frequentist d...       18187   1660   
29484  <p>I think there are several issues (in ascend...       37083   5554   
68668  <p>Since the purpose here is presumably to obt...       16108   1673   
17994  <p>The chi-square test is good as long as the ...       11860   2359   
64739  <p>For independence you want to show that the ...        2196    206   
23134  <p>It is true that each element of a multivari...       17791   5927   
7688   <p>for some reason, people have difficulty gra...        5606    967   
12835  <p>The problem starts with your sentence :</p>...        3000    456   
78546  <p>Any paper that disproves the nil null hypot...         208     28   
43711  <p>Consistency of an estimator means that as t...       22275   7395   

       UpVotes  DownVotes  
9986       230          4  
29484     8641        125  
68668       63          6  
17994      883          7  
64739      298          6  
23134     2122        412  
7688      1547          8  
12835      140          7  
78546       43          0  
43711     2619         42

Analizing the rows that are missing the ViewCount, their features related to the post(CommentCount, Score and Body) are filled. Therefore these data will be important for the analisys so we do not drop them.

In [41]:
posts_from_users.loc[~missing_viewcount,['ViewCount',
                                         'Score',
                                         'CommentCount',
                                         'Reputation',
                                         'Views','UpVotes'
                                         ,'DownVotes']].corr().ViewCount.sort_values(ascending = False)

ViewCount       1.000000
Score           0.532106
Reputation      0.057293
Views           0.056347
UpVotes         0.046272
CommentCount    0.044713
DownVotes       0.033729
Name: ViewCount, dtype: float64

Checking the correlation between ViewCount and the others numeric features, considering rows that do not have NaN values in the ViewCount column, the Score is the one more related(53.21%). So to fill the viewcount 

In [45]:
posts_from_users.loc[~missing_viewcount,:].groupby(by = 'Score', as_index = False).ViewCount.mean()

Score  ViewCount
0      -19      271.0
1      -13      749.0
2      -10      159.0
3       -9      250.0
4       -8      189.0
..     ...        ...
99     124    34780.0
100    152    29229.0
101    156    64481.0
102    184    66071.0
103    192    60964.0

[104 rows x 2 columns]

## Adjust the data types in order to avoid future issues. Which ones should be changed? 

In [34]:
posts_from_users = posts_from_users.drop(index = posts_from_users.loc[missing_viewcount, :].index)

In [35]:
posts_from_users.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 42188 entries, 0 to 90882
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   post_id       42188 non-null  int64  
 1   Score         42188 non-null  int64  
 2   user_id       42188 non-null  float64
 3   ViewCount     42188 non-null  float64
 4   CommentCount  42188 non-null  int64  
 5   Body          42188 non-null  object 
 6   Reputation    42188 non-null  int64  
 7   Views         42188 non-null  int64  
 8   UpVotes       42188 non-null  int64  
 9   DownVotes     42188 non-null  int64  
dtypes: float64(2), int64(7), object(1)
memory usage: 3.5+ MB


Columns user_id and ViewCount should be int.

In [36]:
posts_from_users.user_id.astype('category')
posts_from_users.ViewCount.astype('int64')

0         1278
2        67396
6         1220
9         1022
10        7733
         ...  
90878       17
90879       13
90880       19
90881        9
90882        5
Name: ViewCount, Length: 42188, dtype: int64

# Bonus 1: (filtering) What is the average number of comments for users who are above the average reputation?

*Hint:* Calculate the average of the user Reputation. Store it in a variable called `avg_reputation` and then use that variable for filtering the dataset and generating the results for each case (for the case in which `Reputation > {avg_reputation}` and etc.

*Hint 2:* You could create a variable based on that condition and use the group by function perform the task above.

In [37]:
avg_reputation = posts_from_users['Reputation'].mean()
posts_from_users.loc[posts_from_users['Reputation'] > avg_reputation, :].CommentCount.mean()

2.2574113073060103

# Bonus 2: (grouping) Group your dataframe by the Reputation of your user. Calculate the mean value of ViewCount and CommentCount for each reputation value.

Suppose the missing values on ViewCount are due a systemic error and you wanted to guess what values should have been there in the first place, but the system abended.

Would that be an interesting candidate for inputting the value for the missing `ViewCount` values? If so, input it with these values.

In [39]:
posts_from_users.groupby(by = 'Reputation')[['ViewCount', 'CommentCount']].mean()

ViewCount  CommentCount
Reputation                           
1            123.006787      1.748303
2            438.916667      2.333333
3            164.804706      1.720000
4            230.512821      2.410256
5            321.833333      2.125000
...                 ...           ...
31170       2318.625000      1.375000
37083       1201.666667      2.000000
44152        461.419355      3.645161
65272        570.625000      2.750000
87393       1191.500000      6.750000

[872 rows x 2 columns]

## refs

Sample database used: https://relational.fit.cvut.cz/dataset/Stats

Stack-overflow database: https://www.brentozar.com/archive/2015/10/how-to-download-the-stack-overflow-database-via-bittorrent/
